ui/src/services/csrfTokenApi.ts

```diff
@@ -0,0 +1,125 @@
+import { AxiosInstance, InternalAxiosRequestConfig, isAxiosError } from 'axios';
+import { randomId, showToast } from '@prefecthq/prefect-design';
+import { Api, AxiosInstanceSetupHook, PrefectConfig } from '@prefecthq/prefect-ui-library'
+
+export type CsrfToken = {
+    token: string | null;
+    expiration: Date | null;
+    client: string;
+    issued: Date | null;
+}
+
+export type CsrfTokenResponse = {
+    token: string;
+    expiration: Date;
+}
+
+export class CsrfTokenApi extends Api {
+    public csrfToken: CsrfToken = {
+        token: null,
+        expiration: null,
+        client: randomId(),
+        issued: null,
+    };
+    public csrfSupportEnabled = true;
+    private refreshTimeout: ReturnType<typeof setTimeout> | null = null;
+    private ongoingRefresh: Promise<void> | null = null;
+
+    public constructor(apiConfig: PrefectConfig, instanceSetupHook: AxiosInstanceSetupHook | null = null) {
+        super(apiConfig, instanceSetupHook)
+        this.startBackgroundTokenRefresh()
+    }
+
+    public async addCsrfHeaders(config: InternalAxiosRequestConfig) {
+        if (this.csrfSupportEnabled) {
+            await this.refreshCsrfToken()
+            config.headers = config.headers || {}
+            config.headers['Prefect-Csrf-Token'] = this.csrfToken.token
+            config.headers['Prefect-Csrf-Client'] = this.csrfToken.client
+        }
+    }
+
+    private async refreshCsrfToken(force: boolean = false) {
+        if (!force && !this.shouldRefreshToken()) {
+            return this.ongoingRefresh ?? Promise.resolve();
+        }
+
+        if (this.ongoingRefresh) {
+            return this.ongoingRefresh;
+        }
+
+        const refresh = async () => {
+            try {
+                const response = await this.get<CsrfTokenResponse>(`/csrf-token?client=${this.csrfToken.client}`);
+                const now = new Date();
+                this.csrfToken.token = response.data.token;
+                this.csrfToken.expiration = response.data.expiration;
+                this.csrfToken.issued = now;
```

Comment:
super nit: Since we're not using `now` for anything else I'd just create the date inline. 
```suggestion
                this.csrfToken.issued = new Date();
```